File: ws2812.py
#!/usr/bin/env python

import random
import time
import serial

from clock import Clock
from scheduler import Scheduler
random.seed()

from rndAge import RndAge
from word import Word
from util import  Cmd

def waitAck():
	for w in range(1000):
		if ser.inWaiting()>0:
			data = ser.read(ser.inWaiting())
			break

def writeSerial(buff):
	ser.write(buff)
	waitAck()

def setBg():
	writeSerial(Cmd.setBg().serialize())
	writeSerial(Cmd.flush(0, 5).serialize())

# runner = RndAge()
# runner = Word(["times=1", "Hello, world!"])
runner = Scheduler(RndAge())
runner.prepend_task(Word(["times=1", "Hello, world!"]))
# runner.set_interval(Clock(), 3000)
runner.set_cronjob(Clock(), "0 */5 * * *")

while True:
	try:
		ser = serial.Serial("/dev/ttyS0", baudrate=115200, parity="E")
		setBg()
		prev = time.time()
		while True:
			# print("opening tmp file")
			# with open('/tmp/light', 'r') as f:
			cmds = runner.run()
			if not isinstance(cmds, list):
				runner = RndAge()    # TODO: pipe
			else:
				print("n cmd="+str(len(cmds)))
				for cmd in cmds:
					writeSerial(cmd.serialize())
			now = time.time()
			time.sleep(max(0.1-(now-prev), 0))
			prev = now
	except Exception as ex:
		print(ex)
		time.sleep(5)

File: ws2812.js
const io = require('./io-mock');
const U = require('./utils');
const RndAge = require('./rnd-age');
const Word = require('./word');

function delay(ms) {
    return function(data) {
        return new Promise(function(resolve, reject) {
            setTimeout(resolve, ms, data);
        });
    }
}

// var runner = new RndAge();
var runner = new Word(["times=1", "Hello, world!"]);
function step() {
    function gen(cmd) {
        return function() {
            return io.writeSerial(cmd.serialize());
        };
    }

    var cmds = runner.run();
    if (!Array.isArray(cmds)) {
        io.sendFifo(cmds);
        return Promise.resolve();
    }
    console.log("n cmd=" + cmds.length);
    var promise = Promise.resolve();
    for (var cmd of cmds) {
        // console.log(cmd);
        promise = promise.then(gen(cmd));
    }
    return promise;
}

function loop() {
    return step().then(delay(100)).then(loop);
}

function openAndLoop() {
    io.openSerial()
    .then(loop)
    .catch(function(err) {
        console.log("err: " + err);
        return delay(5000)().then(openAndLoop);
    });
}


function setBg() {
    function gen(i) {
        return function() {
            return io.writeSerial((new U.Cmd(i, 0, U.STRIP_LEN, U.BGCOLOR)).serialize());
        };
    }

    var promise = Promise.resolve();
    for (var i = 0; i < 5; i++) {
        promise = promise.then(gen(i));
    }

    return promise;
}

io.openSerial()
.then(setBg)
.then(io.closeSerial)
.then(openAndLoop);


io.onFifo(function(data) {
    const words = data.split(' ');
    const mode = words[0];
    switch (mode.toLowerCase()) {
        case 'word':
        case 'wordleft':
            runner = new Word(words.slice(1));
            break;
        case 'rndage':
        default:
            runner = new RndAge();
    }
});
io.startFifo('/tmp/pipe');
